src/Language/TokenReaderInterface.php

```diff
@@ -6,20 +6,16 @@
 {
 
     /**
-     * @param int $code
-     * @param int $pos
-     * @param int $line
-     * @param int $col
-     * @param Token $prev
-     * @return Token
+     * @param string $body
+     * @param int    $bodyLength
+     * @param int    $code
+     * @param int    $pos
+     * @param int    $line
+     * @param int    $col
+     * @param Token  $prev
+     * @return Token|null
      */
-    public function read(
-      int $code,
-      int $pos,
-      int $line,
-      int $col,
-      Token $prev
-    ): Token;
+    public function read(string $body, int $bodyLength, int $code, int $pos, int $line, int $col, Token $prev): ?Token;
```

Comment:
Change the signature to accept the `$body` and `$bodyLength` so that we can void calculating the body length multiple times (this can be an expensive operation as we are using `mb_strlen` to do so).

Reply:
Very good. :)